File: C/7576.c
#include <stdio.h>  // ≈‰∏∂≈‰
#include <stdlib.h>

int graph[1000][1000];
int dx[4] = { -1, 1, 0, 0 };
int dy[4] = { 0, 0, -1, 1 };
int cnt = 0;

typedef struct
{
	struct node* next;
	int data_x;
	int data_y;
}node;

typedef struct
{
	node* front;
	node* rear;
	int count;
}queue;

void init_queue(queue* q)
{
	q->count = 0;
	q->front = q->rear = NULL;
}

int is_empty(queue* q)
{
	if (q->count <= 0)
		return 1;
	else
		return 0;
}

void push(queue *q, int x, int y)
{
	node* temp = malloc(sizeof(node));
	temp->data_x = x;
	temp->data_y = y;
	temp->next = NULL;
	if (is_empty(q))
	{
		q->front = temp;
	}
	else
	{
		q->rear->next = temp;
	}
	q->rear = temp;
	q->count++;
}

node pop(queue* q)
{
	if (is_empty(q))
		return;
	else
	{
		node* temp;
		node RETURN;
		temp = q->front;
		RETURN.data_x = temp->data_x;
		RETURN.data_y = temp->data_y;
		q->front = temp->next;
		q->count--;
		free(temp);

		return RETURN;
	}
}

void bfs(queue *q, int m, int n)
{
	int x, y;
	while (!is_empty(q))
	{
		node POP = pop(q);
		x = POP.data_x;
		y = POP.data_y;
		for (int i = 0; i < 4; i++)
		{
			int move_x = x + dx[i];
			int move_y = y + dy[i];
			if (move_x >= 0 && move_y >= 0 && move_x < n && move_y < m)
			{
				if (graph[move_x][move_y] == 0)
				{
					graph[move_x][move_y] = graph[x][y] + 1;
					push(q, move_x, move_y);
					cnt--;
				}
			}
		}
	}

	if (cnt == 0)
		printf("%d", graph[x][y] - 1);
	else
		printf("-1");
}


int main()
{
	int m, n;
	scanf("%d %d", &m, &n);

	queue q;
	init_queue(&q);
	for (int i = 0; i < n; i++)
	{
		for (int j = 0; j < m; j++)
		{
			scanf("%d", &graph[i][j]);
			if (graph[i][j] == 1)
				push(&q, i, j);
			else if (graph[i][j] == 0)
				cnt++;
		}
	}
	
	bfs(&q, m, n);

	return 0;
}

File: JAVA/src/_7576.java
import java.util.LinkedList;
import java.util.Queue;
import java.util.Scanner;

// 토마토
public class _7576 {
    public static int[] dx = {1, -1, 0, 0};
    public static int[] dy = {0, 0, 1, -1};

    public static int bfs(int[][] graph, Queue<int[]> queue, int n, int m) {
        while (!queue.isEmpty()) {
            int[] data = queue.poll();
            for (int i = 0; i < 4; i++) {
                int move_x = data[0] + dx[i];
                int move_y = data[1] + dy[i];
                if (move_x >= 0 && move_y >= 0 && move_x < n && move_y < m) {
                    if (graph[move_x][move_y] == 0) {
                        queue.add(new int[]{move_x, move_y});
                        graph[move_x][move_y] = graph[data[0]][data[1]] + 1;
                    }
                }
            }
        }

        int max = Integer.MIN_VALUE;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                // 안익은 토마토가 존재하는 경우
                if (graph[i][j] == 0) {
                    return -1;
                }

                max = Math.max(max, graph[i][j]);
            }
        }

        if (max == 1) {
            return 0;
        } else {
            return max - 1;
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int m = sc.nextInt();
        int n = sc.nextInt();
        int[][] graph = new int[n][m];
        Queue<int[]> queue = new LinkedList<>();

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                graph[i][j] = sc.nextInt();
                if (graph[i][j] == 1) {
                    queue.add(new int[]{i, j});
                }
            }
        }

        System.out.println(bfs(graph, queue, n, m));

    }
}
